src/stream/src/executor/watermark_filter.rs

```diff
@@ -0,0 +1,477 @@
+    info: ExecutorInfo,
+    table: StateTable<S>,
+}
+
+impl<S: StateStore> WatermarkFilterExecutor<S> {
+    pub fn new(
+        input: BoxedExecutor,
+        watermark_expr: BoxedExpression,
+        event_time_col_idx: usize,
+        ctx: ActorContextRef,
+        table: StateTable<S>,
+    ) -> Self {
+        let info = input.info();
+
+        Self {
+            input,
+            watermark_expr,
+            event_time_col_idx,
+            ctx,
+            info,
+            table,
+        }
+    }
+}
+
+impl<S: StateStore> Executor for WatermarkFilterExecutor<S> {
+    fn execute(self: Box<Self>) -> super::BoxedMessageStream {
+        self.execute_inner().boxed()
+    }
+
+    fn schema(&self) -> &risingwave_common::catalog::Schema {
+        &self.info.schema
+    }
+
+    fn pk_indices(&self) -> super::PkIndicesRef<'_> {
+        &self.info.pk_indices
+    }
+
+    fn identity(&self) -> &str {
+        &self.info.identity
+    }
+}
+
+impl<S: StateStore> WatermarkFilterExecutor<S> {
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(self: Box<Self>) {
+        let Self {
+            input,
+            event_time_col_idx,
+            watermark_expr,
+            ctx,
+            info,
+            mut table,
+        } = *self;
+        let watermark_type = watermark_expr.return_type();
+        assert_eq!(
+            watermark_type,
+            input.schema().data_types()[event_time_col_idx]
+        );
+        let mut input = input.execute();
+
+        let first_barrier = expect_first_barrier(&mut input).await?;
+        table.init_epoch(first_barrier.epoch);
+        // The first barrier message should be propagated.
+        yield Message::Barrier(first_barrier);
+
+        // Initiate and yield the first watermark.
+        let mut current_watermark =
+            Self::get_global_max_watermark(&table, watermark_type.clone()).await?;
+
+        yield Message::Watermark(Watermark::new(
+            event_time_col_idx,
+            current_watermark.clone(),
+        ));
+
+        #[for_await]
+        for msg in input {
+            let msg = msg?;
+            match msg {
+                Message::Chunk(chunk) => {
+                    let chunk = chunk.compact();
+
+                    // Empty chunk should not be processed.
+                    if chunk.cardinality() == 0 {
+                        continue;
+                    }
+
+                    let watermark_array = watermark_expr
+                        .eval_infallible(chunk.data_chunk(), |err| {
+                            ctx.on_compute_error(err, &info.identity)
+                        });
+
+                    // Build the expression to calculate watermark filter.
+                    let watermark_filter_expr = Self::build_watermark_filter_expr(
+                        watermark_type.clone(),
+                        event_time_col_idx,
+                        current_watermark.clone(),
```

Comment:
IIUC, the result will be inaccurate if we use the watermark literal from the last chunk? Should we use the accumulative-max as the right side?